keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -192,584 +173,1331 @@ logs are lost.
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
 
 ### Upgrade / Downgrade Strategy
-When upgrading no changes should be needed to maintain existing behaviour as all of this behaviour is optional and disabled by default.
-To activate the feature they will need to enable the feature gate and mark their containers as sidecars in the container spec.
 
-When downgrading `kubectl`, users will need to remove the sidecar field from any of their Kubernetes manifest files as `kubectl` will refuse to apply manifests with an unknown field (unless you use `--validate=false`).
+#### Upgrade strategy
+
+Existing sidecars (implemented as regular containers) will still work as
+intended, as in the past we don't recognize the new field `restartPolicy` today.
+
+Upgrade will not change any current behaviors.
+
+#### Downgrade strategy
+
+First, there will be no effect on any workload that doesn't use a new field. Any
+combination of feature gate enabled/disabled or version skew will work as usual
+for that workload.
+
+Due to the new field added to `initContainers` to turn them into sidecars,
+downgrading to the version without this feature will not be possible -  the Pod
+spec will be rejected by the control plane and all kubelets during
+unmarshalling. Workloads will have to be deleted and recreated with the old way
+of handling sidecars.
+
+The same problem will arise if the user wants to deploy such workloads on older
+clusters, or clusters without the feature flag enabled.
+
+For the scenarios when the feature gate is disabled on control plane, but not
+disabled on kubelet, users will not be able to schedule Pods with the new field.
+
+For the scenarios when the feature gate is disabled on kubelet, but enabled on
+control plane, users will be able to create these Pods, but kubelet will reject
+those 
```

Comment:
It may reject if it already understands the field and see it.

I guess it will simple treat them as regular init-containers in the older releases, no?

Reply:
treating them as regular Init containers will drive the Pod unusable - initialization will never complete. I have added e2e test case to validate this behavior and make sure we will reject those